File: app/src/main/java/me/ycdev/android/devtools/sampler/mem/MemoryUtils.java
package me.ycdev.android.devtools.sampler.mem;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Debug;

public class MemoryUtils {
    /**
     * Get memory stats of specified processes.
     */
    public static ProcMemStat[] getProcessMemoryStat(Context cxt, int[] pids) {
        ActivityManager am = (ActivityManager) cxt.getSystemService(Context.ACTIVITY_SERVICE);
        final int N = pids.length;
        Debug.MemoryInfo[] pidsMemInfo = am.getProcessMemoryInfo(pids);
        ProcMemStat[] pidsMemStat = new ProcMemStat[N];
        for (int i = 0; i < N; i++) {
            int pid = pids[i];
            Debug.MemoryInfo memInfo = pidsMemInfo[i];
            pidsMemStat[i] = new ProcMemStat(pid);
            pidsMemStat[i].memPss = memInfo.getTotalPss();
            pidsMemStat[i].memPrivate = memInfo.getTotalPrivateDirty();
        }
        return pidsMemStat;
    }
}

File: app/src/main/java/me/ycdev/android/devtools/sampler/mem/MemoryUtils.kt
package me.ycdev.android.devtools.sampler.mem

import android.app.ActivityManager
import android.content.Context

object MemoryUtils {
    /**
     * Get memory stats of specified processes.
     */
    fun getProcessMemoryStat(
        cxt: Context,
        pids: IntArray
    ): Array<ProcMemStat?> {
        val am = cxt.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val pidsMemInfo = am.getProcessMemoryInfo(pids)
        val pidsMemStat = arrayOfNulls<ProcMemStat>(pids.size)
        for (i in pids.indices) {
            val pid = pids[i]
            val memInfo = pidsMemInfo[i]
            pidsMemStat[i] = ProcMemStat(pid)
            pidsMemStat[i]!!.memPss = memInfo.totalPss
            pidsMemStat[i]!!.memPrivate = memInfo.totalPrivateDirty
        }
        return pidsMemStat
    }
}
